service/s3/endpoints.go

```diff
@@ -4325,3 +4239,602 @@ func (r *resolver) ResolveEndpoint(
 	}
 	return endpoint, fmt.Errorf("endpoint rule error, %s", "A region must be set when sending requests to S3.")
 }
+
+type endpointParamsBinder interface {
+	bindEndpointParams(*EndpointParameters)
+}
+
+func bindEndpointParams(input endpointParamsBinder, options Options) *EndpointParameters {
```

Comment:
this does look much simpler. i think when i first implemented it i was trying to follow a pattern common in this SDK of only copying out of `Options` what you need (see my other comment). but seeing how much simpler this is makes me seem to prefer this more.